[jit] AddConstant and findConstant for ClassType

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29220 [quant][graphmode] Enable foldbn tests
* #29219 [jit] Add find_constant in module
* #29218 [jit] Preserve constant from ConcreteModuleType to ClassType
* **#29217 [jit] AddConstant and findConstant for ClassType**

Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D18799955](https://our.internmc.facebook.com/intern/diff/D18799955)

jerryzh168 committed:
[jit] AddConstant and findConstant for ClassType

Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito reviewed (commented):
API looks fine but this will not get serialized correctly as is. Also needs tests. Minimum for this PR is to make sure serialization fails if a class type has constants rather than silently removing them.

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1558,6 +1572,9 @@ struct CAFFE2_API ClassType : public NamedType {
   std::vector<std::string> attributeNames_;
   std::vector<TypePtr> attributeTypes_;
+  // Mapping of constant names -> their value.
+  std::vector<std::string> constantNames_;
```
This creates two namespaces for constants and slots, which allows there to be both a slot `foo` and a constant `foo`. We do not want that, so we need to add checks that prevent these namespaces from overlapping.

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1240,6 +1240,23 @@ struct PythonPrintImpl {
+
+      indent();
+      body_ << name << " : " << "Final[" + v.type()->python_str() << "] = ";
+      if (v.isString()) {
```
Similar here. This introduces a non-complete if-statement branch that will cause a huge number of reported issues in the future. This should use the already existing `printConstant` function.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1240,6 +1240,23 @@ struct PythonPrintImpl {
+      indent();
+      body_ << name << " : " << "Final[" + v.type()->python_str() << "] = ";
+      if (v.isString()) {
+        body_ << "'" << v.toStringRef() << "'\n";
```
string is not quoted

zdevito commented on torch/csrc/jit/script/script_type_parser.cpp:
```diff
@@ -82,6 +82,13 @@ TypePtr ScriptTypeParser::subscriptToType(
     auto key_type = parseTypeFromExpr(subscript.subscript_exprs()[0]);
     auto value_type = parseTypeFromExpr(subscript.subscript_exprs()[1]);
     return DictType::create(key_type, value_type);
+  } else if (typeName == "Final"){
```
This is an incomplete implementation of Final in the parser -- it gives the impression that we support it but then we just ignore it.

jerryzh168 replied:
what should be the proper support here? add Final to the type system?

zdevito replied:
The simplest thing to do would probably be for the constant parser to parse the expect Final before parsing the inner type.

jerryzh168 replied:
do we want to allow Final only when we parse constant or do we want to support it as a general type? we can remove these code if we just want to support it when we parse constant

suo replied:
We can only support `Final` as a class constant annotation today, because we don't have const-ness otherwise in the type system. 

I think the best way to structure this is for `ScriptTypeParser::parseConstant` to become `parseClassConstant(Assign assignment)`, which takes care of parsing the `Final` type expr and parsing the rhs value together. This part should be removed to give a correct error message that `Final` doesn't work generically.

zdevito reviewed (commented):
See inline comments. There are some issues with maintenance, and because this changes serialization and would be hard to fix, these probably should be addressed in this PR.

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
Updated according to @zdevito 's comments

suo commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1240,6 +1240,18 @@ struct PythonPrintImpl {
+      const auto& v = moduleType->getConstant(name).value();
+
+      indent();
+      body_ << name << " : " << "Final[" + v.type()->python_str() << "] = ";
```
I think you should replace the `+` with `<<`

suo reviewed (commented):
Generally looks good to me with a small restructuring. I'd like to see tests that verify that we can save/load with constant information intact, but they probably need to be added later in the stack, when you've exposed constant stuff to `Module`/Python.

zdevito commented on torch/csrc/jit/script/script_type_parser.h:
```diff
@@ -28,6 +28,8 @@ class TORCH_API ScriptTypeParser {
 
   FunctionSchema parseSchemaFromDef(const Def& def, bool skip_self);
 
+  c10::IValue parseConstant(const Expr& type, const Expr& expr);
```
This seems the the wrong place to put this code. First, it doesn't use `this` and this is the type parser, so it should not expose an API that parses values. Maybe make it a separate function to make this clear.

jerryzh168 replied:
It does use `parseBaseTypeName` and `evaluateDefaults` which are part of `ScriptTypeParser` and constant is part of type as of now..maybe I can change the name to `parseClassConstant` to make it more clear?

jerryzh168 replied:
> Generally looks good to me with a small restructuring. I'd like to see tests that verify that we can save/load with constant information intact, but they probably need to be added later in the stack, when you've exposed constant stuff to `Module`/Python.

Yeah, we can add official tests later, I have test locally to make sure this works as well.

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on test/cpp/jit/test_class_type.cpp:
```diff
@@ -28,5 +28,18 @@ void testClassTypeAddRemoveAttr() {
+  auto cls = ClassType::create("foo.bar", cu);
+  cls->addConstant("const1", IValue(1));
+  cls->addConstant("const2", IValue(2));
+  ASSERT_TRUE(cls->numConstants(), 2);
```
`ASSERT_EQ` or use ==. I think this is just testing that 2 is true :p

jamesr66a commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1518,6 +1520,39 @@ struct CAFFE2_API ClassType : public NamedType {
+  c10::optional<IValue> getConstant(const std::string& name) const;
+
+  size_t numConstants() const {
+    AT_ASSERT(constantNames_.size() == constantValues_.size());
```
`TORCH_INTERNAL_ASSERT`, `AT_ASSERT` is deprecated

jamesr66a commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -307,11 +308,12 @@ struct SourceImporterImpl : public Resolver,
                 // This is a regular attribute assignment, of the form:
                 //   foo : Tensor
```
This comment is now stale

jamesr66a commented on torch/csrc/jit/script/script_type_parser.cpp:
```diff
@@ -332,6 +332,28 @@ FunctionSchema ScriptTypeParser::parseSchemaFromDef(
 }
 
+c10::IValue ScriptTypeParser::parseClassConstant(const Assign& assign) {
+  TORCH_INTERNAL_ASSERT(assign.lhs().kind() == TK_VAR);
```
Please `throw ErrorReport` instead of `TORCH_INTERNAL_ASSERT`. People maybe editing the serialized code manually and it would be confusing to see these shown as internal asserts

jamesr66a reviewed (approved):
This looks okay to me. I think the outstanding concern is going to be how constants interact with e.g. device placement in the future, but I think for now this is fine.

@zdevito @suo please chime in if there's anything left blocking this PR

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D18799955](https://our.internmc.facebook.com/intern/diff/D18799955)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] AddConstant and findConstant for ClassType"


Summary:
We want to preserve constant information in ClassType so that
users can access the constants in the module by name.
This is also used later for freezing some attribute(converting
attributes to constant)

Test Plan:
tbd

Reviewers:
mvz
Subscribers:

Tasks:

Tags:

Differential Revision: [D18799955](https://our.internmc.facebook.com/intern/diff/D18799955)

[ghstack-poisoned]
